src/Workspaces/Core/Desktop/Workspace/MSBuild/MSBuildProjectLoader.cs

```diff
@@ -276,17 +280,50 @@ public LoadState(IReadOnlyDictionary<string, ProjectId> projectPathToProjectIdMa
             public void Add(ProjectInfo info)
             {
                 _projectIdToProjectInfoMap.Add(info.Id, info);
-                _projectInfoList.Add(info);
+                //Memoized results of ProjectAlreadyReferencesProject may no longer be correct;
+                //reset the cache.
+                ProjectAlreadyReferencesProjectMemo = new Dictionary<ProjectId, Dictionary<ProjectId, bool>>();
             }
 
-            public bool TryGetValue(ProjectId id, out ProjectInfo info)
+            private bool TryGetValue(ProjectId id, out ProjectInfo info)
             {
                 return _projectIdToProjectInfoMap.TryGetValue(id, out info);
             }
 
-            public IReadOnlyList<ProjectInfo> Projects
+            /// <summary>
+            /// Returns true if the project identified by <paramref name="fromProject"/> has a reference (even indirectly)
+            /// on the project identified by <paramref name="targetProject"/>.
+            /// </summary>
+            public bool ProjectAlreadyReferencesProject(ProjectId fromProject, ProjectId targetProject)
             {
-                get { return _projectInfoList; }
+                Dictionary<ProjectId, bool> fromProjectMemo;
+
+                if ( !ProjectAlreadyReferencesProjectMemo.TryGetValue(fromProject, out fromProjectMemo))
+                {
+                    fromProjectMemo = new Dictionary<ProjectId, bool>();
+                    ProjectAlreadyReferencesProjectMemo.Add(fromProject, fromProjectMemo);
+                }
+
+                bool answer;
+
+                if ( !fromProjectMemo.TryGetValue(targetProject, out answer))
+                {
+                    ProjectInfo info;
+                    answer =
+                        TryGetValue(fromProject, out info) &&
+                        info.ProjectReferences.Any(pr =>
+                            pr.ProjectId == targetProject ||
+                            ProjectAlreadyReferencesProject(pr.ProjectId, targetProject)
+                        );
+                    fromProjectMemo.Add(targetProject, answer);
+                }
+
+                return answer;
+            }
+
+            public IEnumerable<ProjectInfo> Projects
```

Comment:
>IEnumerable [](start = 19, length = 11)

Changing this type from `IReadOnlyList` to `IEnumerable` is a breaking api change